### PR TITLE
[DEV-369] Handle Bitcoin Inputs

### PR DESCRIPTION
**Pull request by Kolibri1990**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/payment/models/crypto-input/crypto-input.entity.ts [hunk 1]
```diff
@@ -7,17 +7,19 @@ import { IEntity } from 'src/shared/models/entity';
...
   CRYPTO_STAKING = 'CryptoStaking',
   BUY_FIAT = 'BuyFiat',
+  BUY_CRYPTO = 'BuyCrypto',
   CRYPTO_CRYPTO = 'CryptoCrypto',
   UNKNOWN = 'Unknown',
 }
 
 @Entity()
-@Index('txAssetRoute', (input: CryptoInput) => [input.inTxId, input.asset, input.route], { unique: true })
+@Index('txAssetRoute', (input: CryptoInput) => [input.inTxId, input.asset, input.route, input.amount], { unique: true })
```
[the remark this answers is not in the document]
Comment (Kolibri1990) [reply src/payment/models/crypto-input/crypto-input.entity.ts:22]: without tx id?

[the remark this answers is not in the document]
Comment (davidleomay) [reply src/payment/models/crypto-input/crypto-input.entity.ts]: Sorry, forgot the comma: TX, vout, asset and route.

File: src/payment/models/crypto-input/crypto-input.entity.ts [hunk 2]
```diff
@@ -7,28 +7,33 @@ import { IEntity } from 'src/shared/models/entity';
...
   CRYPTO_STAKING = 'CryptoStaking',
   BUY_FIAT = 'BuyFiat',
+  BUY_CRYPTO = 'BuyCrypto',
   CRYPTO_CRYPTO = 'CryptoCrypto',
   UNKNOWN = 'Unknown',
 }
 
 @Entity()
-@Index('txAssetRoute', (input: CryptoInput) => [input.inTxId, input.asset, input.route], { unique: true })
+@Index('txAssetRoute', (input: CryptoInput) => [input.inTxId, input.asset, input.route, input.vout], { unique: true })
```
Comment (davidleomay) [inline src/payment/models/crypto-input/crypto-input.entity.ts:22]: Update index name.

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    // get deposit route
+    const route = await this.getDepositRoute(utxo.address);
+    if (!route) {
+      console.error(`Failed to process crypto input. No matching route for ${utxo.address} found.`);
+      return null;
+    }
+
+    return this.cryptoInputRepo.create({
+      inTxId: utxo.txid,
+      outTxId: null, // will be set after crypto forward
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:107]: Can be removed

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      isConfirmed: false,
+      amlCheck: route.user.userData.kycStatus === KycStatus.REJECTED ? AmlCheck.FAIL : AmlCheck.PASS,
+      type: CryptoInputType.BUY_CRYPTO,
+    });
+  }
+
+  private async forwardInputs(): Promise<void> {
+    const inputs = await this.cryptoInputRepo.find({
+      where: {
+        outTxId: null,
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:122]: Use `IsNull()`

File: src/payment/models/crypto-input/defi-input.service.ts [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+
+    // only DFI for staking
+    if (route.type === RouteType.STAKING && assetEntity.name != 'DFI') {
+      console.log(`Ignoring non-DFI crypto input (${amount} ${asset}) on staking route. History entry:`, history);
+      return null;
+    }
+
+    return this.cryptoInputRepo.create({
+      inTxId: history.txid,
+      outTxId: null, // will be set after crypto forward
```
Comment (davidleomay) [inline src/payment/models/crypto-input/defi-input.service.ts:235]: Can be removed

File: src/payment/models/crypto-input/defi-input.service.ts [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+          : route.type === RouteType.STAKING
+          ? CryptoInputType.CRYPTO_STAKING
+          : CryptoInputType.UNKNOWN,
+    });
+  }
+
+  private async forwardInputs(): Promise<void> {
+    const inputs = await this.cryptoInputRepo.find({
+      where: {
+        outTxId: null,
```
Comment (davidleomay) [inline src/payment/models/crypto-input/defi-input.service.ts:258]: Use `IsNull()`

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  private async saveInputs(): Promise<void> {
+    const utxos = await this.btcClient.getUtxo();
+
+    const newInputs = await this.createEntities(utxos).then((i) => i.filter((h) => h != null));
+
+    newInputs.length > 0 && console.log(`New crypto inputs (${newInputs.length}):`, newInputs);
+
+    // side effect, assuming that cryptoStakingRepo and stakingRepo are faultless on save
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:59]: Comment can be removed.

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    // min. deposit
+    if (utxo.amount.toNumber() < Config.node.minBtcDeposit) {
+      console.log(`Ignoring too small crypto input (${utxo.amount.toNumber()} 'BTC'`);
+      return null;
+    }
+
+    // get deposit route
+    const route = await this.getDepositRoute(utxo.address);
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:99]: This does not work. Need to get the CryptoRoute.

File: src/payment/models/crypto-input/crypto-input.service.ts [hunk 9]
```diff
@@ -61,426 +34,49 @@ export class CryptoInputService {
...
-  private async sendFeeUtxo(address: string): Promise<string> {
-    return await this.client.sendUtxo(Config.node.utxoSpenderAddress, address, Config.node.minDfiDeposit / 2);
-  }
-
-  private async getDepositRoute(address: string): Promise<Sell | Staking> {
+  async getDepositRoute(address: string): Promise<Sell | Staking> {
     return (
       (await this.sellService.getSellByAddress(address)) ?? (await this.stakingService.getStakingByAddress(address))
     );
   }
```
Comment (davidleomay) [inline src/payment/models/crypto-input/crypto-input.service.ts:81]: Move to defi-input.service

File: src/payment/models/crypto-input/crypto-input.service.ts [hunk 10]
```diff
@@ -61,426 +34,49 @@ export class CryptoInputService {
...
-      }
-    } catch (e) {
-      console.error('Exception during fee UTXO retrieval:', e);
-    }
-  }
-
   // --- HELPER METHODS --- //
-  private async checkNodeInSync(): Promise<{ headers: number; blocks: number }> {
-    const { blocks, headers } = await this.client.getInfo();
+  async checkNodeInSync(client: NodeClient): Promise<{ headers: number; blocks: number }> {
```
Comment (davidleomay) [inline src/payment/models/crypto-input/crypto-input.service.ts:70]: Mark as protected

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    const inputs = await this.cryptoInputRepo.find({
+      where: {
+        outTxId: null,
+        amlCheck: AmlCheck.PASS,
+        route: { deposit: { blockchain: Blockchain.BITCOIN } },
+      },
+      relations: ['route'],
+    });
+
+    if (inputs.length == 0) return;
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:129]: Re-add the log

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+  // --- CONFIRMATION HANDLING --- //
+  @Interval(300000)
+  async checkConfirmations(): Promise<void> {
+    try {
+      await this.checkNodeInSync(this.btcClient);
+
+      const unconfirmedInputs = await this.cryptoInputRepo.find({
+        select: ['id', 'outTxId', 'isConfirmed'],
+        where: { isConfirmed: false, outTxId: Not(IsNull()) },
+      });
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:154]: Only verify BTC inputs

File: src/payment/models/crypto-input/defi-input.service.ts [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+  // --- CONFIRMATION HANDLING --- //
+  @Interval(300000)
+  async checkConfirmations(): Promise<void> {
+    try {
+      await this.checkNodeInSync(this.client);
+
+      const unconfirmedInputs = await this.cryptoInputRepo.find({
+        select: ['id', 'outTxId', 'isConfirmed'],
+        where: { isConfirmed: false, outTxId: Not(IsNull()) },
+      });
```
Comment (davidleomay) [inline src/payment/models/crypto-input/defi-input.service.ts:310]: Only verify DeFi inputs

File: src/payment/models/crypto-input/crypto-input.service.ts [hunk 14]
```diff
@@ -1,51 +1,24 @@
...
-
-  private client: DeFiClient;
-
   constructor(
-    nodeService: NodeService,
-    private readonly cryptoInputRepo: CryptoInputRepository,
-    private readonly assetService: AssetService,
+    readonly cryptoInputRepo: CryptoInputRepository,
     private readonly sellService: SellService,
     private readonly stakingService: StakingService,
```
Comment (davidleomay) [inline src/payment/models/crypto-input/crypto-input.service.ts:20]: Remove dependency

File: src/ain/node/btc-client.ts [hunk 15]
```diff
@@ -7,25 +7,21 @@ export class BtcClient extends NodeClient {
...
 
-  async send(addressTo: string, txId: string, amount: number, vout: number): Promise<string> {
+  async send(addressTo: string, txId: string, amount: number, vout: number, btcFee: number): Promise<string> {
     return this.callNode(
       (c) =>
         c.call(
           NodeCommand.SEND,
           [
-            [{ [addressTo]: this.roundAmount(amount - this.btcUtxoFee) }],
+            [{ [addressTo]: this.roundAmount(amount - btcFee) }],
```
Comment (davidleomay) [inline src/ain/node/btc-client.ts:16]: Need to multiply btcFee with vByte count. Use e.g. https://bitcoinops.org/en/tools/calc-size/ (single input/output)

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 16]
```diff
@@ -135,7 +137,11 @@ export class BtcInputService extends CryptoInputService {
   }
 
   private async forwardUtxo(input: CryptoInput, address: string): Promise<void> {
-    const outTxId = await this.btcClient.send(address, input.inTxId, input.amount, input.vout);
+    const { fastestFee } = await this.callApi<{ fastestFee: number; halfHourFee: number; hourFee: number }>(
+      this.btcFeeUrl,
+    );
+    const btcFee = fastestFee < Config.crypto.fee * input.amount ? fastestFee : Config.crypto.fee * input.amount;
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:143]: I think this is not related to `Config.crypto.fee`, which is what the user pays. The factor here should be `0.0005% = 0.000005`, the user fee will be much higher.

Code improvement: `const btcFee = Math.min(fastestFee, 0.000005 * input.amount);`

Comment (davidleomay) [reply src/payment/models/crypto-input/btc-input.service.ts:143]: Move the logic to separate method `getFeeRate()` or similar.

File: src/config/config.ts [hunk 17]
```diff
@@ -170,8 +170,10 @@ export class Configuration {
     dexWalletAddress: process.env.DEX_WALLET_ADDRESS,
     outWalletAddress: process.env.OUT_WALLET_ADDRESS,
     stakingWalletAddress: process.env.STAKING_WALLET_ADDRESS,
+    bitcoinWalletAddress: process.env.BITCOIN_WALLET_ADDRESS,
```
Comment (davidleomay) [inline src/config/config.ts:173]: Do we need this address?

File: src/payment/models/crypto-input/btc-input.service.ts [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+      },
+      relations: ['route'],
+    });
+
+    if (inputs.length == 0) return;
+    console.log(`New crypto inputs (${inputs.length}):`, inputs);
+
+    for (const input of inputs) {
+      try {
+        await this.forwardUtxo(input, Config.node.bitcoinWalletAddress);
```
Comment (davidleomay) [inline src/payment/models/crypto-input/btc-input.service.ts:132]: Use DFX's Cake BTC address.
